Clone saved query feature

sfalkoff opened the pull request:
## What does this PR do?

This PR builds off of https://github.com/keen/explorer/pull/180, a PR created by @menismu, which offers users the ability to clone saved queries. This feature was requested in https://github.com/keen/explorer/issues/151. Many thanks to @menismu for implementing this feature!

The updates in this PR include:

- Makes the clone button white.
- A small update to the clone success message copy.
- Uses `_.cloneDeep` to ensure the source query and newly cloned query reference different objects.
- Clones `metadata.visualization.chart_type` to ensure that cloned queries have the original query visualization type set as default. This also resolves an issue where cloned queries were inconsistently displaying the error "A chart type is required. Check out the .type() method." in the UI.
- Adds some additional test coverage for the changes noted above.

![screen shot 2016-12-20 at 12 36 34 pm](https://cloud.githubusercontent.com/assets/6183404/21367264/54165278-c6b2-11e6-8f1b-bab3748ddda7.png)

## How should this be tested?

- Get the branch running locally and run the tests with `npm run test`.
- QA cloning saved queries in the browser. Test several user flows (e.g. clone a clone, update the original query after it has been cloned, delete a clone, etc).

## Related Tickets
https://github.com/keen/explorer/pull/180, #151

joannecheng commented on client/js/app/components/explorer/index.js:
```diff
@@ -82,6 +82,14 @@ var Explorer = React.createClass({
   },
+  
+  cloneQueryClick: function(event) {
+	event.preventDefault();
```
it looks like the whitespace is a little off in this function.

joannecheng commented on client/js/app/stores/ExplorerStore.js:
```diff
@@ -537,6 +538,19 @@ ExplorerStore.dispatchToken = AppDispatcher.register(function(action) {
+    case ExplorerConstants.EXPLORER_CLONE:
+      var source = ExplorerStore.get(action.id);
+      _create({ query: _.cloneDeep(source.query), 
+                metadata : {
```
unnecessary space after `metadata`

joannecheng commented on test/unit/components/explorer/index_spec.js:
```diff
@@ -262,6 +262,47 @@ describe('components/explorer/index', function() {
+        var props = _.assign({}, this.component.props, { persistence: {} });
+        this.component = TestHelpers.renderComponent(Explorer, props);
+      });
+      it('should add a new explorer in the store', function () {
```
this is hard to read. Can we add line breaks before each `it(` statement?

joannecheng commented on test/unit/stores/ExplorerStoreSpec.js:
```diff
@@ -121,6 +121,67 @@ describe('stores/ExplorerStore', function() {
     });
   });
+  
+  describe('clone', function () {
```
extra space after `function`

joannecheng commented on test/unit/stores/ExplorerStoreSpec.js:
```diff
@@ -121,6 +121,67 @@ describe('stores/ExplorerStore', function() {
   });
+  
+  describe('clone', function () {
+  	it('should only clone query data and metadata.visualization.chart_type', function() {
```
Inconsistent whitespace - we're using 2 spaces.

joannecheng commented on test/unit/stores/ExplorerStoreSpec.js:
```diff
@@ -121,6 +121,67 @@ describe('stores/ExplorerStore', function() {
+  	  var source = ExplorerStore.get('abc123');
+  	  ExplorerActions.clone(source.id);
+      var clone = ExplorerStore.getLast();
+  	  var keys = Object.keys(ExplorerStore.getAll());
```
Extra tabs

dustinlarimer commented:
This is working really well! I say :shipit: pending @joannecheng's code review 👍

sfalkoff commented:
@joannecheng I addressed your comments in https://github.com/keen/explorer/pull/193/commits/f393c555b8f887d852fe213fd1e7ef2da9d6f565. Thank you and @dustinlarimer for taking a look!

joannecheng commented:
Everything looks good 🚢

dustinlarimer commented:
🎉 🎉 🎉
